### PR TITLE
Problem: state streamers are not integrated

### PR DESCRIPTION
**Pull request by yihuang**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

Solution:
- integrate the file streamer and add integration test for it.

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -340,6 +346,30 @@
...
+		streamingDir := filepath.Join(cast.ToString(appOpts.Get(flags.FlagHome)), "data", FileStreamerDirectory)
+		if err := os.MkdirAll(streamingDir, os.ModePerm); err != nil {
+			panic(err)
+		}
+
+		// default to exposing all
+		exposeStoreKeys := make([]sdk.StoreKey, 0, len(keys))
+		for _, storeKey := range keys {
+			exposeStoreKeys = append(exposeStoreKeys, storeKey)
+		}
```
[the remark this answers is not in the document]
Comment (yihuang) [reply app/app.go:362]: This non-determinism map iteration is fine I think.

File: integration_tests/test_streamer.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def test_streamers(cronos):
+    """
+    - check the streaming files are created
+    - try to parse the state change sets
+    """
+    # inspect the first state change of the first tx in genesis
+    path = cronos.node_home(0) / "data/file_streamer/block-0-tx-0"
+    _, body, _ = decode_stream_file(open(path, "rb").read())
```
Comment (yihuang) [inline integration_tests/test_streamer.py:59]: The ability of extracting the kv pairs is enough to recover all the historical states.

**Review by mmsqe — COMMENTED** [review]
may I ask if need apply streamingListener.ListenDeliverTx in [v0.46.1](https://github.com/cosmos/cosmos-sdk/blob/v0.46.1/baseapp/abci.go#L281)

File: integration_tests/test_streamer.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    # creation of the validator account
+    assert body[0]["store_key"] == "acc"
```
Comment (mmsqe) [inline integration_tests/test_streamer.py:61]: seems need use object here
```
assert body[0].store_key == "acc"
assert body[0].key == b"\x01" + HexBytes(ADDRS["validator"])
```

File: app/app.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmsqe) [reply app/app.go:362]: not sure if we need just sort this slice

File: integration_tests/test_streamer.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+from cprotobuf import Field, ProtoEntity, decode_primitive
+from hexbytes import HexBytes
+
+from .utils import ADDRS
+
+
+class StoreKVPairs(ProtoEntity):
+    # the store key for the KVStore this pair originates from
+    store_key = Field("string", 1)
+    # true indicates a set operation, false indicates a delete operation
```
Comment (yihuang) [inline integration_tests/test_streamer.py:10]:
```suggestion
    # true indicates a delete operation
```

File: integration_tests/test_streamer.py (hunk 3 above)
Comment (yihuang) [reply integration_tests/test_streamer.py:61]: you are right, but there's another issue right now, the main branch version don't dump the states for genesis txs.

Comment (mmsqe) [reply integration_tests/test_streamer.py:61]: you mean miss implement ListenDeliverTx in [DeliverTx](https://github.com/cosmos/cosmos-sdk/blob/9c51d9ac9824ce29e99601dbbc74317c1f4f5fdb/docs/architecture/adr-038-state-listening.md) in main?

Comment (yihuang) [reply integration_tests/test_streamer.py:61]: yes

Comment (yihuang) [reply integration_tests/test_streamer.py:61]: I think it get deleted accidentally here: https://github.com/cosmos/cosmos-sdk/pull/11979

File: app/app.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomtau) [reply app/app.go:362]: it's definitely safe that it's not used in the consensus code in that sense -- not sure if `NewStreamingService` expects those store keys to always be in the same order (e.g. when restarted) though

[the remark this answers is not in the document]
Comment (adu-crypto) [reply app/app.go:362]: it should be ok as the `storeKeys` is consumed by `NewStreamingService` like this:
```go
// NewStreamingService creates a new StreamingService for the provided writeDir, (optional) filePrefix, and storeKeys
func NewStreamingService(writeDir, filePrefix string, storeKeys []types.StoreKey, c codec.BinaryCodec) (*StreamingService, error) {
	listenChan := make(chan []byte)
	iw := NewIntermediateWriter(listenChan)
	listener := types.NewStoreKVPairWriteListener(iw, c)
	listeners := make(map[types.StoreKey][]types.WriteListener, len(storeKeys))
	// in this case, we are using the same listener for each Store
	for _, key := range storeKeys {
		listeners[key] = append(listeners[key], listener)
	}
...
```
these store keys are used as the keys of the mapping

File: go.mod [hunk 5]
```diff
@@ -184,7 +186,7 @@ require (
 )
 
 replace (
-	github.com/cosmos/cosmos-sdk => github.com/cosmos/cosmos-sdk v0.46.1
+	github.com/cosmos/cosmos-sdk => github.com/yihuang/cosmos-sdk v0.43.0-beta1.0.20220920021715-149d0d651552
```
Comment (yihuang) [inline go.mod:189]: manually backport https://github.com/cosmos/cosmos-sdk/pull/13334 to 0.46.

**Comment by yihuang** [comment]
> may I ask if need apply streamingListener.ListenDeliverTx in [v0.46.1](https://github.com/cosmos/cosmos-sdk/blob/v0.46.1/baseapp/abci.go#L281)

yeah, exactly, just created the PR: https://github.com/cosmos/cosmos-sdk/pull/13334

**Review by mmsqe — APPROVED** [review]

**Pull request merged**